internal/testutils/daemon.go

```diff
@@ -0,0 +1,134 @@
+package testutils
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+type daemonOptions struct {
+	cachePath  string
+	cacheDB    string
+	socketPath string
+}
+
+// DaemonOption represents an optional function that can be used to override some of the daemon default values.
+type DaemonOption func(*daemonOptions)
+
+// WithCachePath overrides the default cache path of the daemon.
+func WithCachePath(path string) DaemonOption {
+	return func(o *daemonOptions) {
+		o.cachePath = path
+	}
+}
+
+// WithCacheDB overrides the default cache database of the daemon.
+func WithCacheDB(db string) DaemonOption {
+	return func(o *daemonOptions) {
+		o.cacheDB = db
+	}
+}
+
+// WithSocketPath overrides the default socket path of the daemon.
+func WithSocketPath(path string) DaemonOption {
+	return func(o *daemonOptions) {
+		o.socketPath = path
+	}
+}
+
+// RunDaemon runs the daemon in a separate process and returns the socket path and a channel that will be closed when
+// the daemon stops.
+func RunDaemon(ctx context.Context, t *testing.T, execPath string, args ...DaemonOption) (socketPath string, stopped chan struct{}) {
+	t.Helper()
+
+	opts := &daemonOptions{}
+	for _, opt := range args {
+		opt(opts)
+	}
+
+	// Socket name has a maximum size, so we can't use t.TempDir() directly.
+	tempDir, err := os.MkdirTemp("", "authd-daemon4tests")
+	require.NoError(t, err, "Setup: failed to create temp dir for tests")
+	t.Cleanup(func() { os.RemoveAll(tempDir) })
+
+	if opts.cachePath == "" {
+		opts.cachePath = filepath.Join(tempDir, "cache")
+		require.NoError(t, os.MkdirAll(opts.cachePath, 0700), "Setup: failed to create cache dir")
+	}
+
+	if opts.cacheDB != "" {
+		CreateDBFromYAML(t, filepath.Join("testdata", "db", opts.cacheDB+".db.yaml"), opts.cachePath)
+	}
+
+	if opts.socketPath == "" {
+		opts.socketPath = filepath.Join(tempDir, "authd.socket")
```

Comment:
Can you please make `tempDir` or actually the socket name more random? Just a few random digits would be ok... As having it in `/tmp/authd.sock` all the times may block cases in which you've already something there already (like myself actually having the daemon running).

In general I'd keep the place for all these dirs in a disposable place though that is cleaned up when needed.

Reply:
Err, here I see the tmpdir is fine since that's the one generated by `authd-daemon4tests` but there's somewhere in the code that still uses `/tmp/authd.sock` as reflected by the golden files.

Now, I see that we need to have a stable path for that in order to be predictable. But maybe just make it a bit more custom (i.e `/tmp/authd-pam-cli-tests.sock`).

Reply:
Or... Even better, maybe you can just use a random value, but... When you're comparing the results you replace the one in the golden files? It may need a different way to generate them though.

But ideally one should be able to run these tests in parallel.

Reply:
I see... Well, the way would probably make all the recording phase to be programmed, so that the generation of the tapes is not done by hand. And they can be regenerated. But well, probably not worth for now.

Reply:
We don't have a way to program the recording. My first idea was to have the tapes recorded manually, but (at least in Go) if you try to run vhs as a subprocess (i.e. `exec.Command`), it will spawn yet another one, e.g:
- Main Go process
  |----> `exec.Command("vhs", tape)` // We only have access to the stdin of this one.
             |----> `vhs` process // We needed access to the stdin of this subprocess to control the execution 
                                                    and generate tapes "dynamically"